### DISCUSSION TITLE
Need guidance in setting up production server

### DISCUSSION THREAD

**Discussion by fahadmalikawan** [exadel-inc/CompreFace#1067] [Q&A]
I've installed the library on Ubuntu Server. I needed the most accurate algorithm, so I chose Subcenter Arcface, without GPU.
I installed the system on docker.

The machine specs are:
Intel Core-i7 13700k
Corsair 64GB 5600Mhz
2 x Samsung 980 Pro 2TB NVME (Speed upto 7000 MBPS each)

Still I get 4-5 seconds per image recognition.
I really need help to optimize the system.
I'm wondering what GPU should I use for best efficiency
Have I implemented the system in correct way or not.
Should I implement on kubernetes (Don't know how)
What should be the settings in .env file etc.

Any guidance/help would be appritiated.
Thanks

**Comment by pospielov** [comment]
Subcenter Arcface is a very compute-consuming mode. 
4-5 seconds on the CPU is quite a good performance. To improve it, you can use a GPU. 
With my GPU 2070 on my laptop, I achieved about 7 requests per second. 
Of course, if you get a more powerful GPU, it will be even better.

One thing to consider - Subcenter Arcface model requires a little bit more than 4Gb of GPU memory. So I would recommend getting GPU with at least 6Gb of memory. 
In .env there is a parameter uwsgi_processes that says how many processes to use. For each process, it will take memory, both RAM and GPU. So e.g. if you have 6Gb GPU, you should set it to 1, if you have a 12Gb GPU you should set it to 2, If you have a 24Gb GPU, you should set it to 5. 
We have repository with kubernetes configuration:
https://github.com/exadel-inc/compreface-kubernetes
Its advantage is that you can easily add more nodes to speed up the recognition, but adding GPU gives more advantages. So use it only if even adding GPU is not enough. Or if you want to build a more reliable system.

**Comment by fahadmalikawan** [comment]
@pospielov Thanks for the guidance.
I proceeded further meanwhile and just wanted to share my experience, so it may benefit others who are struggling.
I have tested the system with SubCenter Arcface with the following instructions
1- 
Ryzen 5500 Desktop with 32GB RAM. Docker on Windows. Didn't modify the default .env settings
Its performacne was good, recognizing one face within a second. However on idle, the CPU usage goes high after sometime by WSL-2 (Seems like WSL bug)
RAM goes upto 25GB

2-
Ryzen 5625U Laptop with 16 GB RAM. Docker on Windows. Didn't modify the default .env settings
Recognition was fast. around a second per face. RAM is not enough for docker

3-
Intel Core-i7 13700k
RAM 64GB 5600Mhz
Fastest NVME
Tested on Windows with Ubuntu Server both
My expectation was very high this time but results were strange.
5-8 seconds per face. No matter how i tried.

Maybe Intel is not good for this algorithm and Ryzen is good. (No clear idea)
Then I tried the GPU version
a-
RTX 3060Ti 8GB With default .env settings i.e. 2 uWSGI processes. 
Each process eats up more or little less than 4GB. so mostly throws error by full memory consumption

RTX 3060Ti 8GB with 1 uWSGI process.
Results are amazing. Many faces in 1 seconds. Can't remember how many. No memory leakage. No issues, great performance

RTX 4070 12GB
System failed to provide results. Throwing error:
error during synchronization between servers: unexpected end of file from server executing post http://compreface-core:3000/find_faces
Tried hard, reinstalled everything. OS, drivers, Nvidia Runtime etc. No luck
I guess, Nvidia 4xxx series is not yet supported, if it does for someone, please guide here

RTX 3090 -> 5uWSGI process
Outclass performance, No issues. A picture of football crowd with 74 faces detected and recognized withing 800ms

**Reply by pospielov** [reply]
Nvidia 4xxx are not supported yet, we are working on this.

**Reply by doctorttt** [reply]
When do you think Nvidia 4000 series will be supported? I have a 4060ti and not working in Compreface.

**Reply by pospielov** [reply]
We are almost finished with the next release. But we still need some time for the last fixes and make regression. 
Hope we will release in July

**Reply by doctorttt** [reply]
Thanks for the update. Looking forward to the next release and use my Nvidia 4000 in Compreface.
